bip-0054/test_vectors/README.md

```diff
@@ -0,0 +1,102 @@
+## BIP54 test vectors
+
+This folder contains a set of test vectors for each mitigation introduced in the BIP. This document
+presents them in more detail.
+
+The code used to generate half of the test vectors is included with the implementation and available
+[here][other-vectors]. The other half requires mining mainnet blocks and is [published
+separately][bip54-miner]. In both cases it is implemented as a regular Bitcoin Core unit test, and
+the test vectors are persisted as a JSON file if the `UPDATE_JSON_TESTS` preprocessor directive is
+set (off by default).
+
+To compile the [header][header-miner] and [block][block-miner] miners you may have to link to
+libatomic explicitly. This can be achieved like so:
+```
+cmake -B atomicbuild -DAPPEND_LDFLAGS="-latomic"
+cmake --build atomicbuild/ -j $(nproc)
+```
+
+[Premined headers][premined-headers] are also provided along with the header miner to allow changing
+some of the last headers without having to re-generate the whole chain.
+
+
+### Difficulty adjustment exploits
+
+The [`timestamps.json`](./timestamps.json) test vectors exercise the two constraints on block header
+timestamps introduced by BIP54 to mitigate the Timewarp and Murch-Zawy attacks. Each test case
+features a chain of mainnet headers starting from the genesis block, and whether this header chain
+is valid by BIP54 rules. Each test case also contains a comment describing why this particular chain
+is (in)valid according to BIP54.  All test cases are valid according to current Bitcoin consensus
+rules. It is intended to be used to test a BIP54 implementation by feeding the header chain to a
+Bitcoin node implementation, enforcing the BIP54 rules on this chain from genesis.
+
+The test vector file features a JSON array of JSON objects, each corresponding to a test case. Each
+JSON object features the following entries:
+- `header_chain`: a JSON array of strings. An ordered list of hex-encoded mainnet block headers.
+- `valid`: a JSON boolean. Whether this chain of headers is valid according to BIP54.
+- `comment`: a JSON string. Description of the test case.
+
+For the purpose of testing a Timewarp fix, a Timewarp attack was included early on in the history of
+testnet3. An implementer of BIP54 may want to ensure that syncing testnet3 by enforcing BIP54 since
+genesis will treat block `00000000118da1e2165a19307b86f87eba814845e8a0f99734dce279ca3fb029` as
+invalid.
```

Comment:
Very open question: how hard it would be to coordinate a Murch-Zawy attack variant on testnet3.